### PR TITLE
return static instead of self when static context detected

### PR DESCRIPTION
**Pull request by orklah**
This PR propose improving the handling of functions that return $this.

The goal is to transform this code:
```php
class HelloWorld
{
    public function sayHello()
    {
        return $this;
    }
}
```
into:

```php
class HelloWorld
{
    /**
     * @return $this
     */
    public function sayHello(): self
    {
        return $this;
    }
}
```

Bonus points: I want to propose another PR so that this example return `static` instead of `self` in PHP8 but I'm not sure how I'm supposed to access to the php version inside an Atomic. Can you give me a hint?

### CODE REVIEW

File: src/Psalm/Type.php [hunk 1]
```diff
@@ -110,9 +110,13 @@ public static function getStringFromFQCLN(
         ?string $namespace,
         array $aliased_classes,
         ?string $this_class,
-        bool $allow_self = false
+        bool $allow_self = false,
+        bool $was_static = false
     ) : string {
         if ($allow_self && $value === $this_class) {
+            if ($was_static) {
```
Comment (staabm) [inline src/Psalm/Type.php:117]: The condition should be inverse?

**Comment by muglug** [comment]
I'd prefer `@return static`, especially as it's PHP-8-compatible

**Comment by orklah** [comment]
> I'd prefer `@return static`, especially as it's PHP-8-compatible

Oh? I thought `$this` was more narrow than `static` (probably just semantic, `$this`means that the object itself was returned when `static` just meant the type was the same as the object being called on, but not necessary the same instance) but `static` is fine by me too.

File: src/Psalm/Type.php (hunk 1 above)
Comment (orklah) [reply src/Psalm/Type.php:117]: Why do you think so? This boolean is true when in static context so we add `$this`/`static` on the phpdoc when it's true

**Comment by muglug** [comment]
Thanks! Psalm treats `$this` and `static` the same in docblocks, FWIW.

**Pull request merged**